libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs

```diff
@@ -894,6 +895,114 @@ public virtual async Task<TokenStatus[]> GetTokenStatusAsync(ITurnContext contex
+                    Bot = activity.Recipient,       // Activity is from the user to the bot
+                    ChannelId = activity.ChannelId,
+                    Conversation = activity.Conversation,
+                    ServiceUrl = activity.ServiceUrl,
+                    User = activity.From,
+                },
+                MsAppId = appId,
+            };
+
+            var serializedState = JsonConvert.SerializeObject(tokenExchangeState);
+            var encodedState = Encoding.UTF8.GetBytes(serializedState);
+            var state = Convert.ToBase64String(encodedState);
+
+            var client = await CreateOAuthApiClientAsync(turnContext).ConfigureAwait(false);
+
+            return await client.BotSignIn.GetSignInResourceAsync(state, null, null, finalRedirect, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Performs a token exchange operation such as for single sign-on.
+        /// </summary>
+        /// <param name="turnContext">Context for the current turn of conversation with the user.</param>
+        /// <param name="connectionName">Name of the auth connection to use.</param>
+        /// <param name="userId">The user id associated with the token..</param>
+        /// <param name="exchangeRequest">The exchange request details, either a token to exchange or a uri to exchange.</param>
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects
+        /// or threads to receive notice of cancellation.</param>
+        /// <returns>If the task completes, the exchanged token is returned.</returns>
+        public virtual async Task<TokenResponse> ExchangeTokenAsync(ITurnContext turnContext, string connectionName, string userId, TokenExchangeRequest exchangeRequest, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var client = await CreateOAuthApiClientAsync(turnContext).ConfigureAwait(false);
+            var result = await client.UserToken.ExchangeAsyncAsync(userId, connectionName, turnContext.Activity.ChannelId, exchangeRequest, cancellationToken).ConfigureAwait(false);
+
+            if (result is ErrorResponse errorResponse)
+            {
+                result = new TokenResponse();
+            }
+
+            if (result is TokenResponse tokenResponse)
+            {
+                // todo: temporary
+                if (string.IsNullOrEmpty(tokenResponse.Token))
+                {
+                    if (!string.IsNullOrEmpty(exchangeRequest.Uri))
+                    {
+                        tokenResponse.Token = "exchangeableToken" + Guid.NewGuid().ToString().Replace("-", string.Empty);
+                    }
+                    else
+                    {
+                        tokenResponse.Token = "token" + Guid.NewGuid().ToString().Replace("-", string.Empty);
+                    }
+                }
+
+                return tokenResponse;
+            }
```

Comment:
Food for thought: I already worry about these methods being part of the Bot Framework Adapter. However, now we are getting to the point where a significant % of the adapter code is signin related. I wonder if we should 1) refactor the signin code out to a class, and have the adapter expose the key public methods and call the satellite class, or 2) refactor the signin code to a class, have it implement IUserTokenProvider + ITokenExchangeProvider, and have a property in adapter that exposes its ITokenExchangeProvider, and finally OAuthPrompt consumes this class. That way we can clean up the adapter. Just for thought.

Reply:
I like the second option...can we really do it? isn't it "brekaing" at this point?

---
In reply to: [375493100](https://github.com/microsoft/botbuilder-dotnet/pull/3320#discussion_r375493100) [](ancestors = 375493100)